Add WebCodec's AudioData cocoa support.

jyavenard opened the pull request:
#### 676d1e61c613c793bdad642eaf70dfa6e408d96c
<pre>
Add WebCodec&apos;s AudioData cocoa support.
<a href="https://bugs.webkit.org/show_bug.cgi?id=283901">https://bugs.webkit.org/show_bug.cgi?id=283901</a>
<a href="https://rdar.apple.com/140782143">rdar://140782143</a>

Reviewed by Youenn Fablet.

It supports conversion from and to all WebCodec&apos;s supported AudioFormat
in both interleaved and planar layout.

Covered by web platform tests which are no longer expected to fail.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/webcodecs/audioData-serialization-expected.txt: Test is now passing.
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-insertable-streams/VideoTrackGenerator.worker-expected.txt: Test no longer fails as it relied on AudioData
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/transfering.https.any-expected.txt: Test fails due to webkit.org/b/284017
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/transfering.https.any.worker-expected.txt: Test fails due to webkit.org/b/284017
* LayoutTests/platform/glib/http/wpt/webcodecs/audioData-serialization-expected.txt: Removed.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
(WebCore::isAudioSampleFormatInterleaved): Don&apos;t pass enum value by const reference.
(WebCore::computeBytesPerSample): Don&apos;t pass enum value by const reference.
(WebCore::computeCopyElementCount): Update as per latest spec. Don&apos;t limit conversion to only float32 with cocoa implementation
(WebCore::audioSampleElementFormat):
(WebCore::audioElementSpan): Utility method to convert original span&lt;uint8_t&gt; to one matching the underlying datatype.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/AudioStreamDescription.h: Add support for U8 and Int24 samples.
* Source/WebCore/platform/audio/PlatformRawAudioData.cpp:
* Source/WebCore/platform/audio/PlatformRawAudioData.h:
(WebCore::PlatformRawAudioData::isGStreamer const): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
(WebCore::AudioSampleBufferList::applyGain):
(WebCore::mixBuffers):
* Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.cpp:
(WebCore::CAAudioStreamDescription::CAAudioStreamDescription):
(WebCore::CAAudioStreamDescription::format const):
* Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.h:
* Source/WebCore/platform/audio/cocoa/CARingBuffer.h:
(WebCore::CARingBuffer::fetchModeForMixing):
* Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp: Added.
(WebCore::PlatformRawAudioData::create):
(WebCore::audioSampleFormatToPCMFormat):
(WebCore::interleavedFormat):
(WebCore::createSampleBuffer):
(WebCore::PlatformRawAudioDataCocoa::PlatformRawAudioDataCocoa):
(WebCore::PlatformRawAudioDataCocoa::asbd const):
(WebCore::PlatformRawAudioDataCocoa::format const):
(WebCore::PlatformRawAudioDataCocoa::sampleRate const):
(WebCore::PlatformRawAudioDataCocoa::numberOfChannels const):
(WebCore::PlatformRawAudioDataCocoa::numberOfFrames const):
(WebCore::PlatformRawAudioDataCocoa::duration const):
(WebCore::PlatformRawAudioDataCocoa::timestamp const):
(WebCore::PlatformRawAudioDataCocoa::memoryCost const):
(WebCore::PlatformRawAudioDataCocoa::isInterleaved const):
(WebCore::planesOfSamples):
(WebCore::PlatformRawAudioData::copyTo):
(WebCore::PlatformRawAudioDataCocoa::sampleBuffer const):
* Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.h: Added.
(isType):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.h: Add bufferAsSpan utility method.
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
(WebCore::PlatformRawAudioData::create):
(WebCore::PlatformRawAudioDataGStreamer::numberOfFrames const):
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.h:
(isType):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
(WebCore::clipAudioBufferList):
* Tools/TestWebKitAPI/Tests/WebCore/CARingBufferTest.cpp:
(TestWebKitAPI::CARingBufferTest::setup):

Canonical link: <a href="https://commits.webkit.org/287332@main">https://commits.webkit.org/287332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/465c74d04393a2602987c29495118e5fd4021a9b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79230 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32602 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83855 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30397 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81363 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67350 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6528 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/61981 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19879 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82297 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52038 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72191 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42287 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49383 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26340 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28785 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70495 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26770 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85247 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6528 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4523 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70226 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6691 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68063 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69474 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17318 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13526 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12346 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6485 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6407 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9874 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8199 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
```diff
@@ -77,6 +78,11 @@ static std::pair<GstAudioFormat, GstAudioLayout> convertAudioSampleFormatToGStre
 
+Ref<PlatformRawAudioData> PlatformRawAudioData::create(Ref<MediaSample>&& sample)
+{
+    return PlatformRawAudioDataGStreamer::create(GRefPtr { sample->platformSample() });
```
```suggestion
    return PlatformRawAudioDataGStreamer::create(GRefPtr { sample->platformSample() });
    return PlatformRawAudioDataGStreamer::create(GRefPtr(sample->platformSample().sample.gstSample));
```

jyavenard replied:
yeah, I had made that change on my linux box, but it didn't make it back to the Mac :(

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
```diff
@@ -75,7 +75,7 @@ bool isAudioSampleFormatInterleaved(const AudioSampleFormat& format)
 }
 
-size_t computeBytesPerSample(const AudioSampleFormat& format)
+size_t computeBytesPerSample(AudioSampleFormat format)
```
```suggestion
// https://w3c.github.io/webcodecs/#compute-copy-element-count
size_t computeBytesPerSample(AudioSampleFormat format)
```
Maybe we should rename the function to computeCopyElementCountInternal?

jyavenard replied:
I'm not sure. This comment doesn't apply to that method.
bytePerSample has little to do with computeCopyElementCount

And computeCopyElementCount is the spec name, and it does exactly the same computation as per the spec

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
```diff
@@ -101,42 +101,97 @@ ExceptionOr<size_t> computeCopyElementCount(const WebCodecsAudioData& data, cons
     auto destFormat = options.format.value_or(platformData->format());
 
+    // 3. If destFormat describes an interleaved AudioSampleFormat and options.planeIndex is greater than 0, throw a RangeError.
     auto isInterleaved = isAudioSampleFormatInterleaved(destFormat);
```
```suggestion
    bool isInterleaved = isAudioSampleFormatInterleaved(destFormat);
```

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.h:
```diff
@@ -30,14 +30,22 @@
+#include <span>
+#include <variant>
+
+
```
Remove one line?

youennf commented on Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp:
```diff
@@ -0,0 +1,320 @@
+{
+    CMAudioFormatDescriptionRef rawFormatDescription;
+    if (PAL::CMAudioFormatDescriptionCreate(kCFAllocatorDefault, &description.streamDescription(), 0, nullptr, 0, nullptr, nullptr, &rawFormatDescription))
+        return nullptr;
```
We could debug log these error cases.

youennf commented on Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp:
```diff
@@ -0,0 +1,320 @@
+    WebAudioBufferList inputList = inputDescription;
+    WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    uint8_t* data = const_cast<uint8_t*>(sourceData.data());
+    size_t sizePlane = numberOfFrames * inputDescription.bytesPerFrame();
```
Is there a risk of overflow?

jyavenard replied:
> Is there a risk of overflow?

no. The checks are performed in `computeCopyElementCount` and we throw if any overflow in the calculation occur before being called here. And the fact that we could create the sourceData in the first place (which is even wider) indicates that there's no overflow.

I could add an assert.

Or maybe do add a check just to make it explicitly clear to the reader that it can't happen

youennf commented on Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp:
```diff
@@ -0,0 +1,320 @@
+    }
+    WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+
+    RetainPtr sample = createSampleBuffer(inputDescription, PAL::CMTimeMake(timestamp, 1000000), numberOfFrames, &inputList);
```
Why passing inputList as a pointer?

jyavenard replied:
hmm, in a previous iteration of the code the list was optional, but it now longer is. So I'll make it a reference. good spotting

youennf commented on Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp:
```diff
@@ -0,0 +1,320 @@
+    return isAudioSampleFormatInterleaved(format) ? CAAudioStreamDescription::IsInterleaved::Yes : CAAudioStreamDescription::IsInterleaved::No;
+}
+
+static RetainPtr<CMSampleBufferRef> createSampleBuffer(const CAAudioStreamDescription& description, const CMTime& time, size_t numberOfFrames, WebAudioBufferList* list)
```
```suggestion
static RetainPtr<CMSampleBufferRef> createSampleBuffer(const CAAudioStreamDescription& description, const CMTime& time, size_t numberOfFrames, const WebAudioBufferList* list)
```
?

jyavenard replied:
`const` here doesn't make the list const itself; just the pointer. But I see what you mean. If we had the object const itself, The CM method would then need a const_cast which is far from great

youennf commented on Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.h:
```diff
@@ -0,0 +1,76 @@
+    size_t numberOfChannels() const final;
+    size_t numberOfFrames() const final;
+    std::optional<uint64_t> duration() const final;
+    int64_t timestamp() const final;
```
It would help if either name or type of duration and timestamp would tell us this is in microseconds.

jyavenard replied:
`timestamp` is the name used in every WebCodec related classes. maybe in a future change; or better yet, make it a MediaTime then there's no confusion and it can handle rational. same for duration

youennf commented on Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp:
```diff
@@ -0,0 +1,320 @@
+            // Simplest case.
+            ASSERT(!planeIndex);
+            size_t frameOffsetInBytes = frameOffset.value_or(0) * audioData.m_description.bytesPerFrame();
+            auto source = unsafeMakeSpan(static_cast<uint8_t*>(sourceList.buffer(0)->mData) + frameOffsetInBytes, sourceList.buffer(0)->mDataByteSize - frameOffsetInBytes);
```
Are we sure we validate enough all values here at the caller site?
Ditto below.

Also, WebAudioBufferList could provide a `span<const uint8_t> bufferAsSpan(int)` so that we would use subspace here.
Or WebAudioBufferList would provide a copy method itself.

jyavenard replied:
The validation is done in Compute Copy Element Count ; I triple checked that code, but dealing with flat raw memory buffer there's always a risk of something in there. The spec itself doesn't make it safe.

But the copyElementCount account for the destination size. I'll try to make the flow a bit less convoluted and clear that it's safe

youennf commented on Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp:
```diff
@@ -0,0 +1,320 @@
+    auto source = planesOfSamples(sourceFormat, sourceList, frameOffset.value_or(0) * (audioData.isInterleaved() ? numberOfChannels() : 1));
+
+    if (!audioData.isInterleaved() && destinationIsInterleaved) {
+        // copy of all channels of the source into the destination buffer and deinterleave.
```
```suggestion
        // Copy all channels of the source into the destination buffer and deinterleave.
```

jyavenard committed:
Add WebCodec's AudioData cocoa support.
https://bugs.webkit.org/show_bug.cgi?id=283901
rdar://140782143

Reviewed by Youenn Fablet.

It supports conversion from and to all WebCodec's supported AudioFormat
in both interleaved and planar layout.

Covered by web platform tests which are no longer expected to fail.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/webcodecs/audioData-serialization-expected.txt: Test is now passing.
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-insertable-streams/VideoTrackGenerator.worker-expected.txt: Test no longer fails as it relied on AudioData
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/transfering.https.any-expected.txt: Test fails due to webkit.org/b/284017
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/transfering.https.any.worker-expected.txt: Test fails due to webkit.org/b/284017
* LayoutTests/platform/glib/http/wpt/webcodecs/audioData-serialization-expected.txt: Removed.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
(WebCore::isAudioSampleFormatInterleaved): Don't pass enum value by const reference.
(WebCore::computeBytesPerSample): Don't pass enum value by const reference.
(WebCore::computeCopyElementCount): Update as per latest spec. Don't limit conversion to only float32 with cocoa implementation
(WebCore::audioSampleElementFormat):
(WebCore::audioElementSpan): Utility method to convert original span<uint8_t> to one matching the underlying datatype.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/AudioStreamDescription.h: Add support for U8 and Int24 samples.
* Source/WebCore/platform/audio/PlatformRawAudioData.cpp:
* Source/WebCore/platform/audio/PlatformRawAudioData.h:
(WebCore::PlatformRawAudioData::isGStreamer const): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
(WebCore::AudioSampleBufferList::applyGain):
(WebCore::mixBuffers):
* Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.cpp:
(WebCore::CAAudioStreamDescription::CAAudioStreamDescription):
(WebCore::CAAudioStreamDescription::format const):
* Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.h:
* Source/WebCore/platform/audio/cocoa/CARingBuffer.h:
(WebCore::CARingBuffer::fetchModeForMixing):
* Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp: Added.
(WebCore::PlatformRawAudioData::create):
(WebCore::audioSampleFormatToPCMFormat):
(WebCore::interleavedFormat):
(WebCore::createSampleBuffer):
(WebCore::PlatformRawAudioDataCocoa::PlatformRawAudioDataCocoa):
(WebCore::PlatformRawAudioDataCocoa::asbd const):
(WebCore::PlatformRawAudioDataCocoa::format const):
(WebCore::PlatformRawAudioDataCocoa::sampleRate const):
(WebCore::PlatformRawAudioDataCocoa::numberOfChannels const):
(WebCore::PlatformRawAudioDataCocoa::numberOfFrames const):
(WebCore::PlatformRawAudioDataCocoa::duration const):
(WebCore::PlatformRawAudioDataCocoa::timestamp const):
(WebCore::PlatformRawAudioDataCocoa::memoryCost const):
(WebCore::PlatformRawAudioDataCocoa::isInterleaved const):
(WebCore::planesOfSamples):
(WebCore::PlatformRawAudioData::copyTo):
(WebCore::PlatformRawAudioDataCocoa::sampleBuffer const):
* Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.h: Added.
(isType):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.h: Add bufferAsSpan utility method.
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
(WebCore::PlatformRawAudioData::create):
(WebCore::PlatformRawAudioDataGStreamer::numberOfFrames const):
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.h:
(isType):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
(WebCore::clipAudioBufferList):
* Tools/TestWebKitAPI/Tests/WebCore/CARingBufferTest.cpp:
(TestWebKitAPI::CARingBufferTest::setup):

Canonical link: https://commits.webkit.org/287332@main
